src/probability/stopping.lean

```diff
@@ -1291,4 +1292,197 @@ by { ext x, rw stopped_value, by_cases hx : x ∈ s; simp [hx] }
 
 end piecewise_const
 
+section hitting
+
+/-- Hitting time: given a stochastic process `u` and a set `s`, `hitting u s n m` is the first time
+`u` is in `s` after time `n` and before time `m` (if `u` does not hit `s` after time `n` and
+before `m` then the hitting time is simply `m`).
+
+The hitting time is a stopping time if the process is adapted and discrete. -/
+noncomputable def hitting [preorder ι] [has_Inf ι] (u : ι → α → β) (s : set β) (n m : ι) : α → ι :=
+λ x, if ∃ j ∈ set.Icc n m, u j x ∈ s then Inf (set.Icc n m ∩ {i : ι | u i x ∈ s}) else m
+
+lemma hitting_of_lt [conditionally_complete_lattice ι]
+  {u : ι → α → β} {s : set β} {n m : ι} {x : α} (h : m < n) :
+  hitting u s n m x = m :=
+begin
+  simp_rw [hitting],
+  have h_not : ¬∃ (j : ι) (H : j ∈ set.Icc n m), u j x ∈ s,
+  { push_neg,
+    intro j,
+    rw set.Icc_eq_empty_of_lt h,
+    simp only [set.mem_empty_eq, forall_false_left], },
+  simp only [h_not, if_false],
+end
+
+lemma hitting_le [conditionally_complete_linear_order ι]
+  {u : ι → α → β} {s : set β} {n m : ι} (x : α) :
+  hitting u s n m x ≤ m :=
+begin
+  cases le_or_lt n m with h_le h_lt,
+  { simp only [hitting],
+    split_ifs,
+    { obtain ⟨j, hj₁, hj₂⟩ := h,
+      have hj_mem : j ∈ set.Icc n m ∩ {i : ι | u i x ∈ s} := set.mem_inter hj₁ hj₂,
+      rw set.mem_Icc at hj₁,
+      refine (cInf_le _ hj_mem).trans hj₁.2,
+      refine bdd_below.inter_of_left _,
+      exact bdd_below_Icc, },
+    { exact le_rfl }, },
+  { rw hitting_of_lt h_lt, },
+
+end
+
+lemma le_hitting [conditionally_complete_lattice ι]
+  {u : ι → α → β} {s : set β} {n m : ι} (x : α) (hmn : n ≤ m) :
+  n ≤ hitting u s n m x :=
+begin
+  simp only [hitting],
+  split_ifs,
+  { obtain ⟨j, hj, hmem⟩ := h,
+    refine le_cInf ⟨j, hj, hmem⟩ _,
+    rintro k ⟨hk₁, hk₂⟩,
+    exact hk₁.1 },
+  { assumption }
+end
+
+section complete_linear_order
+
+variables [complete_linear_order ι] {u : ι → α → β} {s : set β} {f : filtration ι m}
+
+@[simp]
+lemma hitting_eq_Inf (x : α) : hitting u s ⊥ ⊤ x = Inf {i : ι | u i x ∈ s} :=
```

Comment:
I don't think that should be a simp lemma. We will have a lot of results about hitting times, which we cannot use on the Inf.
```suggestion
lemma hitting_eq_Inf (x : α) : hitting u s ⊥ ⊤ x = Inf {i : ι | u i x ∈ s} :=
```